### ISSUE TITLE
Can't annotate the gingiva

### ISSUE THREAD

**Issue by dangis-lll** [Tai-Hsien/MeshSegNet#19]
Hello, I tried to test the model in the middle of machine learning. But I don't know why I didn't separate label 0.
After being treated with pygco, there is no label 0 in the VTP file.
![图片](https://user-images.githubusercontent.com/71537001/127642840-081aca97-c543-494d-a16d-9f631922c753.png)
![图片](https://user-images.githubusercontent.com/71537001/127642862-2f6db319-55c6-428e-a594-5723f6a0d3e5.png)

**Comment by SivHMA** [comment]
Hi!
I have a very easy way to solve this problem, which is to change gingiva'。s label to another number, such as 18 instead of 0.

**Comment by Tai-Hsien** [comment]
Hello,

Does the well-trained model have the same problem? If not, it means the learning pattern is that the model learns rough position of teeth first without gingiva, and then fine-tune boundary and assign the non-teeth-portion as gingiva ultimately.

Best,

Bruce Wu

**Comment by dangis-lll** [comment]
> 
> 
> Hi!
> I have a very easy way to solve this problem, which is to change gingiva's label to another number, such as 18 instead of 0.

Thank you, I'll try this way.

**Comment by dangis-lll** [comment]
> 
> 
> Hello,
> 
> Does the well-trained model have the same problem? If not, it means the learning pattern is that the model learns rough position of teeth first without gingiva, and then fine-tune boundary and assign the non-teeth-portion as gingiva ultimately.
> 
> Best,
> 
> Bruce Wu

No, I haven't finish the train, I'll wait it and see if it will get better.
Thank you for you work.

**Comment by donproc** [comment]
Thank you for the code sharing @Tai-Hsien 
I found the similar issue after i trained model with default lr 1e-3 and 200 epochs. @dangis-lll Have you solved the problem? Do i need try bigger lr? Thanks!

**Comment by Tai-Hsien** [comment]
Hello @donproc 
Interesting. We used to get good results around 200 epochs. Before figuring out the issue, could you keep training it to see what will happen around 400 epochs?

BTW, what is the number of cells of samples during training?

**Comment by donproc** [comment]
> 

@Tai-Hsien I trained 200 epochs with lr 1e-2 and 1e-3, finally i got similar results. I will try more epcohs. By the way, i trained with around 800 cases and different cases have different classes, for example, case 1 has 15 classes and case 2 has 17 classes.

**Comment by dangis-lll** [comment]
> 
Hello,@donproc I have solved this problom.
On the 104 line of the Mesh_dataset.py, `if num_positive > self.patch_size:`, if the number of cells of your samples is greater than self.patch_size, the gingiva's CellId will be deleted, so you need to change the dataset to preserve gingiva.
Hope this will help you.

**Comment by donproc** [comment]
> > 
> 
> Hello,@donproc I have solved this problom. On the 104 line of the Mesh_dataset.py, `if num_positive > self.patch_size:`, if the number of cells of your samples is greater than self.patch_size, the gingiva's CellId will be deleted, so you need to change the dataset to preserve gingiva. Hope this will help you.

Thanks for your help! @dangis-lll I will try you method.

**Comment by donproc** [comment]
> > 
> 
> Hello,@donproc I have solved this problom. On the 104 line of the Mesh_dataset.py, `if num_positive > self.patch_size:`, if the number of cells of your samples is greater than self.patch_size, the gingiva's CellId will be deleted, so you need to change the dataset to preserve gingiva. Hope this will help you.

I tried with 30% negative samples(1800 in 6000), the result does not improve after around 80 epochs. Do you have any recomendation on negative ratio? @Tai-Hsien @dangis-lll

**Comment by Tai-Hsien** [comment]
Hello @donproc,

I know the network will recognize all teeth first; after certain epochs, it will start learning gingiva.

**Comment by donproc** [comment]
> 

Thank you! @Tai-Hsien I will check the result after 100 120 150 175 200 epochs.

**Comment by Tai-Hsien** [comment]
Hello @donproc 

No problem. I'm also curious and would like to figure it out if it still cannot recognize gingiva. Please give me update!

**Comment by donproc** [comment]
> Hello @donproc
> 
> No problem. I'm also curious and would like to figure it out if it still cannot recognize gingiva. Please give me update!

Sure, I will let you know. 
Result after 80 epochs.
![image](https://user-images.githubusercontent.com/16092281/150261760-7d0169d1-fa01-46ec-aac5-58d4ebf3a394.png)
Golden
![image](https://user-images.githubusercontent.com/16092281/150261831-2ebd4f9e-ca33-481b-9d7b-821212ce2dd6.png)

**Comment by donproc** [comment]
@Tai-Hsien I think i made some mistake. I will let you know final result.

**Comment by donproc** [comment]
> Hello @donproc
> 
> No problem. I'm also curious and would like to figure it out if it still cannot recognize gingiva. Please give me update!

It works! The algorithm can learn gingiva.
